### PR TITLE
[rachio] Initial contribution

### PR DESCRIPTION
**Pull request by markus7017**
This binding integrates Rachio sprinkler controllers. The binding uses the Rachio Cloud API. Event callback can be configured to utilized myopenhab.org as a proxy.

### CODE REVIEW

**Comment by robnielsen** [comment]
@markus7017, I'm trying out the rachio binding, and noticed some messages getting logged with INFO. These will need to be changed to debug or higher.

File: bundles/org.openhab.binding.rachio/README.md [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    
+    // Zones
+    Thing rachio:zone:1:XXXXXXXXXXXX-1 "Rachio zone 1" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-2 "Rachio zone 2" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-3 "Rachio zone 3" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-4 "Rachio zone 4" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-5 "Rachio zone 5" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-6 "Rachio zone 6" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-7 "Rachio zone 7" @ "Sprinkler"
+    Thing rachio:zone:1:XXXXXXXXXXXX-8 "Rachio zone 8" @ "Sprinkler"
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/README.md:134]: I'm seeing the following in openhab.log with my 8 zone rachio:
```
2021-04-13 12:47:27.271 [INFO ] [el.core.internal.ModelRepositoryImpl] - Validation issues found in configuration model 'rachio.things', using it anyway:
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>
Provide a thing type ID and a thing ID in this format:
 <thingTypeId> <thingId>

```

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+/**
+ * The {@link RachioDiscoveryService} discovers all devices/zones reported by the Rachio Cloud. This requires the api
+ * key to get access to the cloud data.
+ *
+ * @author Markus Michels - Initial contribution
+ */
+@NonNullByDefault
+public class RachioDiscoveryService extends AbstractDiscoveryService implements DiscoveryService, ThingHandlerService {
+
+    private static final int DISCOVERY_REFRESH_SEC = 900;
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:50]: Why is this necessary every 15 minutes? Once a controller is configured why does this need to continually occur?

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                    RachioZone zone = ze.getValue();
+                    logger.debug("Checking zone with ID '{}'", zone.id);
+
+                    // register thing if it not already exists
+                    ThingUID zoneThingUID = new ThingUID(THING_TYPE_ZONE, bridgeUID, zone.getThingID());
+                    zone.setUID(devThingUID, zoneThingUID);
+                    logger.info("Zone#{} '{}' (id={}) added, enabled={}", zone.zoneNumber, zone.name, zone.id,
+                            zone.getEnabled());
+
+                    if (zone.getEnabled() == OnOffType.ON) {
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:166]: Shouldn't all zones be added, even if they are currently be disabled?  The zones can be ignored in the inbox if needed.

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+                HashMap<String, RachioZone> zoneList = dev.getZones();
+                logger.debug("Found {} zones for this device.", zoneList.size());
+                for (HashMap.Entry<String, RachioZone> ze : zoneList.entrySet()) {
+                    RachioZone zone = ze.getValue();
+                    logger.debug("Checking zone with ID '{}'", zone.id);
+
+                    // register thing if it not already exists
+                    ThingUID zoneThingUID = new ThingUID(THING_TYPE_ZONE, bridgeUID, zone.getThingID());
+                    zone.setUID(devThingUID, zoneThingUID);
+                    logger.info("Zone#{} '{}' (id={}) added, enabled={}", zone.zoneNumber, zone.name, zone.id,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:163]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+                    if (zone.getEnabled() == OnOffType.ON) {
+                        @SuppressWarnings({ "unchecked", "rawtypes" })
+                        Map<String, Object> zproperties = (Map) zone.fillProperties();
+                        DiscoveryResult zoneDiscoveryResult = DiscoveryResultBuilder.create(zoneThingUID)
+                                .withProperties(zproperties).withBridge(bridgeUID)
+                                .withLabel(dev.name + "[" + zone.zoneNumber + "]: " + zone.name).build();
+                        thingDiscovered(zoneDiscoveryResult);
+                    } else {
+                        logger.info("Zone#{} '{}' is disabled, skip thing creation", zone.name, zone.id);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:174]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+                    }
+                }
+            }
+            logger.info("{}  Rachio device initialized.", deviceList.size());
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:178]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            }
+            logger.debug("Found {} devices.", deviceList.size());
+            for (HashMap.Entry<String, RachioDevice> de : deviceList.entrySet()) {
+                RachioDevice dev = de.getValue();
+                logger.debug("Check Rachio device with ID '{}'", dev.id);
+
+                // register thing if it not already exists
+                ThingUID devThingUID = new ThingUID(THING_TYPE_DEVICE, bridgeUID, dev.getThingID());
+                dev.setUID(bridgeUID, devThingUID);
+                logger.info(" Rachio device discovered: '{}' (id {}), S/N={}, MAC={}", dev.name, dev.id,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:144]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                        dev.serialNumber, dev.macAddress);
+                logger.debug("  latitude={}, longitude={}", dev.latitude, dev.longitude);
+                logger.info("   device status={}, paused/sleep={}, on={}", dev.status, dev.getSleepMode(),
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:147]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+    public boolean webhookEvent(RachioEventGsonDTO event) {
+        boolean update = false;
+        try {
+
+            String zoneName = event.zoneName;
+            String evt = event.subType.isEmpty() ? event.type : event.subType;
+            zone.setEvent(evt, getTimestamp()); // and funnel all zone events to the device
+            if (event.type.equals("ZONE_STATUS")) {
+                if (event.zoneRunStatus.state.equals("ZONE_STARTED")) {
+                    logger.info("{}: Zone {} STARTED watering ({}).", thingId, zoneName, event.timestamp);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java:212]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+                    updateChannel(CHANNEL_ZONE_RUN, OnOffType.ON);
+                } else if (event.subType.equals("ZONE_STOPPED") || event.subType.equals("ZONE_COMPLETED")) {
+                    logger.info(
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java:215]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+                            "{}: Zoned {} STOPPED watering (timestamp={}, current={}, duration={}sec/{}min, flowVolume={}).",
+                            thingId, zoneName, event.timestamp, event.zoneCurrent, event.duration,
+                            event.durationInMinutes, event.flowVolume);
+                    updateChannel(CHANNEL_ZONE_RUN, OnOffType.OFF);
+                } else {
+                    logger.info("{}: Event for zone {}: {} (status={}, duration = {}sec)", thingId, event.zoneName,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java:221]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+                            event.summary, event.zoneRunStatus.state, event.duration);
+                }
+                update = true;
+            } else if (event.subType.equals("ZONE_DELTA")) {
+                logger.info("{}: DELTA Event for zone {}: {}.{}", thingId, zone.name, event.category, event.action);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioZoneHandler.java:226]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+                // Set DeviceUID for all zones
+                HashMap<String, RachioZone> zoneList = dev.getZones();
+                for (HashMap.Entry<String, RachioZone> ze : zoneList.entrySet()) {
+                    RachioZone zone = ze.getValue();
+                    ThingUID zoneThingUID = new ThingUID(THING_TYPE_ZONE, bridgeThing.getUID(), zone.getThingID());
+                    zone.setUID(dev.getUID(), zoneThingUID);
+                }
+            }
+
+            logger.info("RachioCloud: Connector initialized");
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java:121]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+                logger.debug("RachioCloud: Bridge is ONLINE");
+                updateStatus(ThingStatus.ONLINE);
+            }
+
+            HashMap<String, RachioDevice> checkDevList = checkApi.getDevices();
+            for (HashMap.Entry<String, RachioDevice> de : checkDevList.entrySet()) {
+                RachioDevice checkDev = de.getValue();
+                RachioDevice dev = deviceList.get(checkDev.id);
+                if (dev == null) {
+                    logger.info("RachioCloud: New device detected: {} - {}", checkDev.id, checkDev.name);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java:219]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+        } finally {
+            if (!errorMessage.isEmpty()) {
+                logger.debug("RachioBridge: {}", errorMessage);
+            }
+            jobPending = false;
+        }
+    }
+
+    public void shutdown() {
+        logger.info("RachioCloud: Shutting down");
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java:267]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+                    logger.debug("Command value is no StringType: {}", command);
+                }
+            } else if (channel.equals(RachioBindingConstants.CHANNEL_DEVICE_RUN)) {
+                if (command == OnOffType.ON) {
+                    logger.debug("START watering zones '{}' ('' = ALL)", dev.getRunZones());
+                    handler.runMultipleZones(dev.getAllRunZonesJson(handler.getDefaultRuntime()));
+                }
+            } else if (channel.equals(RachioBindingConstants.CHANNEL_DEVICE_STOP)) {
+                if (command == OnOffType.ON) {
+                    logger.info("STOP watering for device '{}'", dev.name);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:179]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+                    handler.stopWatering(dev.id);
+                    updateState(RachioBindingConstants.CHANNEL_DEVICE_STOP, OnOffType.OFF);
+                }
+            } else if (channel.equals(RachioBindingConstants.CHANNEL_DEVICE_RAIN_DELAY)) {
+                if (command instanceof DecimalType) {
+                    logger.info("Start rain delay cycle for {} sec", command.toString());
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:185]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java [hunk 18]
```diff
@@ -0,0 +1,580 @@
...
+                        checkApi.getLastApiResult().rateReset);
+                logger.debug("{}", errorCritical);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, errorCritical); // shutdown
+                                                                                                         // bridge+devices+zones
+                return;
+            }
+            if (checkApi.getLastApiResult().isRateLimitWarning()) {
+                skipCalls++;
+                if (skipCalls % RACHIO_RATE_SKIP_CALLS > 0) {
+                    logger.info("RachioCloud: API limit is getting critical -> skip update ({} / {})", skipCalls,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioBridgeHandler.java:204]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+
+            String evt = event.subType.isEmpty() ? event.type : event.subType;
+            dev.setEvent(evt, getTimestamp()); // and funnel all zone events to the device
+            if (etype.equals("DEVICE_STATUS")) {
+                // sub types:
+                // COLD_REBOOT, ONLINE, OFFLINE, OFFLINE_NOTIFICATION, SLEEP_MODE_ON, SLEEP_MODE_OFF, BROWNOUT_VALVE
+                // RAIN_SENSOR_DETECTION_ON, RAIN_SENSOR_DETECTION_OFF, RAIN_DELAY_ON, RAIN_DELAY_OFF
+                logger.debug("Device {} ('{}') changed to status '{}'.", d.name, d.id, event.subType);
+                if (event.subType.equals("COLD_REBOOT")) {
+                    logger.info("{}: Device {}  was restarted, ip={}/{}, gw={}, dns={}/{}, wifi rssi={}.", thingId,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:323]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+                            d.name, d.network.ip, d.network.nm, d.network.gw, d.network.dns1, d.network.dns2,
+                            d.network.rssi);
+                    if (event.network != null) {
+                        dev.setNetwork(event.network);
+                    }
+                } else if (event.subType.equals("ONLINE")) {
+                    logger.info("{}: Device is ONLINE.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:330]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+                    dev.setStatus(event.subType);
+                } else if (event.subType.equals("OFFLINE") || event.subType.equals("OFFLINE_NOTIFICATION")) {
+                    logger.info("{}: Device is OFFLINE (subType = '{}').", thingId, event.subType);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:333]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+                    dev.setStatus(event.subType);
+                } else if (event.subType.equals("SLEEP_MODE_ON")) {
+                    logger.info("{}: Device switch to sleep mode.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:336]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+                    dev.setSleepMode(event.subType);
+                } else if (event.subType.equals("SLEEP_MODE_OFF")) {
+                    logger.info("{}: Device was resumed (exit from sleep mode).", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:339]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+                    dev.setSleepMode(event.subType);
+                } else if (event.subType.equals("RAIN_DELAY_ON")) {
+                    logger.info("{}: Device reporterd Rain Delay ON.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:342]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+                    update = false; // details missing
+                } else if (event.subType.equals("RAIN_DELAY_OFF")) {
+                    logger.info("{}: Device reporterd Rain Delay OFF.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:345]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 26]
```diff
@@ -0,0 +1,390 @@
...
+                    update = false; // details missing
+                } else if (event.subType.equals("RAIN_SENSOR_DETECTION_ON")) {
+                    logger.info("{}: Device reporterd a Rain Sensor ON.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:348]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 27]
```diff
@@ -0,0 +1,390 @@
...
+                    update = false; // details missing
+                } else if (event.subType.equals("RAIN_SENSOR_DETECTION_ON")) {
+                    logger.info("{}: Device reporterd Rain Sensor OFF.", thingId);
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:351]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java [hunk 28]
```diff
@@ -0,0 +1,390 @@
...
+                    update = false; // details missing
+                } else {
+                    update = false; // details missing
+                }
+            } else if (event.type.equals("SCHEDULE_STATUS")) {
+                logger.info("{}: Status {} for schedule {}: {} (start={}, end={}, duration={}min)", thingId,
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/handler/RachioDeviceHandler.java:357]: logger.debug

File: bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+    protected void startBackgroundDiscovery() {
+        logger.debug("Starting background discovery for new Rachio controllers");
+
+        ScheduledFuture<?> discoveryJob = this.discoveryJob;
+        if (discoveryJob == null || discoveryJob.isCancelled()) {
+            discoveryJob = scheduler.scheduleWithFixedDelay(this::discover, 10, DISCOVERY_REFRESH_SEC,
+                    TimeUnit.SECONDS);
+        }
+    }
+
```
Comment (robnielsen) [inline bundles/org.openhab.binding.rachio/src/main/java/org/openhab/binding/rachio/internal/discovery/RachioDiscoveryService.java:109]: Missing:

```
    @Override
    protected void stopBackgroundDiscovery() {
        logger.debug("Stopping background discovery for new Rachio controllers");
        if (discoveryJob != null) {
            discoveryJob.cancel(true);
            discoveryJob = null;
        }
    }
```

**Comment by kaikreuzer** [comment]
@markus7017 Did you see @robnielsen's review comments?

**Comment by fwolter** [comment]
@markus7017 Did you see @kaikreuzer's question if you saw @robnielsen comments?

**Review by jlaur — CHANGES_REQUESTED** [review]
Setting PR status according to pending review comments.

**Comment by lolodomo** [comment]
Marked as "stale", last activity was 14 months ago and there were no answers to review comments.

**Comment by markus7017** [comment]
Hi gusy, sorry. I set other priorities, but just some people asked for OH 3.3 support so I'll work on this shortly

File: bundles/org.openhab.binding.rachio/pom.xml [hunk 30]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.1.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.rachio/pom.xml:10]:
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.rachio/pom.xml [hunk 31]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (jlaur) [inline bundles/org.openhab.binding.rachio/pom.xml:3]:
```suggestion
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
```

See #11833

File: bundles/org.openhab.binding.rachio/pom.xml [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.4.0-SNAPSHOT</version>
```
Comment (wborn) [inline bundles/org.openhab.binding.rachio/pom.xml:10]:
```suggestion
    <version>4.0.0-SNAPSHOT</version>
```

File: bundles/org.openhab.binding.rachio/src/main/resources/OH-INF/binding/binding.xml [hunk 33]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (lsiepel) [inline bundles/org.openhab.binding.rachio/src/main/resources/OH-INF/binding/binding.xml:1]: This needs to be adapted to the new `addon.xml`, the PR to adapt the developer docs is not merged yet, so if you need more details, the corresponding issue is here: https://github.com/openhab/openhab-core/issues/2058
But you can also update your branch and look at another binding's `addon.xml` and adapt it to this initial contribution.

**Comment by lsiepel** [comment]
This PR is 2,5 years old. In https://github.com/openhab/openhab-addons/pull/10434#issuecomment-1179388515 @markus7017 (PR author) commented that his priorities have changed. Now (over a year later) the status quo has not changed. As this PR needs some work (review, fixes etc), i don't think some one else will bring it to a mergable state.
If ever needed it can be re-opened at any time.

**Pull request closed without merging**